### PR TITLE
feat(button): add v5 color support

### PR DESCRIPTION
**Pull request by tihuan**
---
title: "type_of_change(scope_of_work): ticket title"
reviewers: "sds-eng, sds-design"
---

## Summary

**Structural Element (Base, Gene, DNA, Chromosome or Cell)**
Shortcut ticket: [sh-XXXX](link)

This ticket adds MUI button v5 color support, which expands v4's options "primary", "secondary", to `color?: "primary" | "secondary" | "success" | "error" | "warning" | "info";`

Implementation inspiration from: https://github.com/mui-org/material-ui/issues/13875#issuecomment-841869537

demo

## Checklist

- [ ] Default Story in Storybook
- [ ] LivePreview Story in Storybook
- [ ] Test Story in Storybook
- [ ] Tests written
- [ ] Variables from `defaultTheme.ts` used wherever possible
- [ ] If updating an existing component, depreciate flag has been used where necessary
- [ ] Chromatic build verified by @sds-design

### CODE REVIEW

File: src/core/Button/style.ts [hunk 1]
```diff
@@ -8,16 +8,44 @@ import {
...
+   * Currently we're extending MUIv4 Button's `color` props from "primary" and
+   * "secondary" to the following ones.
+   */
+  color?: "primary" | "secondary" | "success" | "error" | "warning" | "info";
+}
+
+type V5ButtonProps = Props & ExtraProps & Omit<ButtonProps, "color">;
+
+// Please keep this in sync with the props used in `ExtraProps`
+const doNotForwardProps = [
```
Comment (jfoo1984) [inline src/core/Button/style.ts:27]: perhaps we could rename this to `nonMuiProps` or something along those lines?  That is a little more expressive around why we don't want to forward these props.  (I'm happy to make this change but wanted to run it by you first).

**Comment by jfoo1984** [comment]
One general comment (and I'm happy to make this change), for the "external" interface for a button (provided in `src/core/Button/index.tsx`), could we exclude `color` and `variant`?  I think that in the `switch` statement, we're basically using `sdsStyle` and `sdsType` to determine what to pass for those values into `styles.ts`.  By exposing them externally, we're possibly allowing someone to use different combinations for those that haven't been specified in storybook (and it seems like one of the major benefits of having a design system is to restrict the variants that are used in the app).  Any thoughts on that?

**Comment by SDS-Design** [comment]
WTF happened here?
________________________________

**Comment by tihuan** [comment]
> WTF happened here?

Woah yeah 🤯 How did you get the notification? Did we accidentally tag you somehow??

**Comment by tihuan** [comment]
> I like this approach and how the styles function more clearly encapsulate different styles and allow us to reuse the function from one place! Thank you for putting this together!
> 
> One general comment (and I'm happy to make this change), for the "external" interface for a button (provided in `src/core/Button/index.tsx`), could we exclude `color` and `variant`? I think that in the `switch` statement, we're basically using `sdsStyle` and `sdsType` to determine what to pass for those values into `styles.ts`. By exposing them externally, we're possibly allowing someone to use different combinations for those that haven't been specified in Storybook (and it seems like one of the major benefits of having a design system is to restrict the variants that are used in the app). Any thoughts on that?

Thanks so much for the thoughtful comment 🤩 🙏 ! 

Yeah, exposing MUI props at the top level is intentional, because our SDS components need to provide the flexibility (and escape hatches) for any product to overwrite the style as needed. The reason is because SDS aims to provide reasonable defaults as recommendations (instead of prescriptions), but ultimately the product team should retain the control of how they want to style their components. This will continue to be needed as external partners start using SDS as well 🙆‍♂️

**Comment by SDS-Design** [comment]
My friend, I have here the full thread and you named it with SDS and sds design, that are my initials, Sandra Diana Schöner. I'm programmer but... Not long, since 3 years and you name a function after me, WTF 🤣.  Thank you for that honor 🙏.

It's angular, I understand it, I think one day I opened an issue and yeah. Here we are.

SDS
________________________________

**Comment by tihuan** [comment]
@SDS-Design That's amazing! Really nice meeting you too, Sandra 😂 🎉 Hopefully you get to use this component library down the road 🤩 👏 

Have an amazing day!

> My friend, I have here the full thread and you named it with SDS and sds design, that are my initials, Sandra Diana Schöner. I'm programmer but... Not long, since 3 years and you name a function after me, WTF 🤣. Thank you for that honor 🙏. It's angular, I understand it, I think one day I opened an issue and yeah. Here we are. SDS
> […](#)

File: src/core/Button/style.ts [hunk 2]
```diff
@@ -8,16 +8,33 @@ import {
   Props,
 } from "../styles";
 
-const sdsPropNames = ["isAllCaps", "isRounded", "sdsStyle", "sdsType"];
+export interface ExtraProps {
+  isAllCaps?: boolean;
+  isRounded?: boolean;
+  /**
+   * TODO(185930): Remove custom `color` prop when we upgrade to MUIv5.
```
Comment (tihuan) [inline src/core/Button/style.ts:15]: @mdjmoore I forgot we haven't merged the button PR 😂 

Also created a ticket already!

https://app.shortcut.com/sci-design-system/story/185930/material-ui-mui-v5-todos

Thank you!

**Review by mdjmoore — APPROVED** [review]
┗(＾0＾)┓

**Comment by tihuan** [comment]
I'll double check this PR addresses specs from https://app.shortcut.com/sci-design-system/story/188176/implement-error-variant-on-button-component !

**Comment by tihuan** [comment]
No longer needed, since MUIv5 migration is done!

**Pull request closed without merging**